src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs

```diff
@@ -0,0 +1,846 @@
+                    0.0022699646742388863, 0.0019091376570554135, 0.0011570531254881296, 0.000697019955113331, 0.00041737721863073713,
+                    0.000248438820613534, 0.00014700521929794912, 8.647365841055832e-05, 5.056939088336744e-05, 2.9400808653120604e-05,
+                    1.6994687082728674e-05, 9.767061541798089e-06
+                };
+
+            private readonly double _threshold;
+            private readonly double _sensitivity;
+            private readonly SrCnnDetectMode _detectMode;
+
+            //used in all modes
+            private double[] _predictArray;
+            private double[] _backAddArray;
+            private double[] _fftRe;
+            private double[] _fftIm;
+            private double[] _magList;
+            private double[] _magLogList;
+            private double[] _spectralList;
+            private double[] _transRe;
+            private double[] _transIm;
+            private double[] _ifftRe;
+            private double[] _ifftIm;
+            private double[] _ifftMagList;
+            private double[] _cumSumList;
+            private double[] _cumSumShift;
+            //used in AnomalyAndMargin mode
+            private double[] _units;
+            private double[] _val;
+            private double[] _ans;
+            private double[] _curWindow;
+
+            public SrCnnEntireModeler(double threshold, double sensitivity, SrCnnDetectMode detectMode)
+            {
+                _threshold = threshold;
+                _sensitivity = sensitivity;
+                _detectMode = detectMode;
+            }
+
+            public void Train(double[] values, ref double[][] results)
+            {
+                if (results == null)
+                {
+                    results = new double[values.Length][];
+                    for (int i = 0; i < results.Length; ++i)
+                    {
+                        results[i] = new double[_outputLengthArray[(int)_detectMode]];
+                    }
+                }
+                else if (results.Length > values.Length)
+                {
+                    Array.Resize<double[]>(ref results, values.Length);
+                }
+                SpectralResidual(values, results, _threshold);
+                //Optional Steps
+                if (_detectMode == SrCnnDetectMode.AnomalyAndMargin)
+                {
+                    GetMargin(values, results, _sensitivity);
+                }
+                else if (_detectMode == SrCnnDetectMode.AnomalyAndExpectedValue)
+                {
+                    GetExpectedValue(values, results);
+                }
+            }
+
+            private void AllocateDoubleArray(ref double[] arr, int length)
+            {
+                if (arr == null)
+                {
+                    arr = new double[length];
+                }
+                else if (arr.Length != length)
```

Comment:
>arr.Length != length [](start = 25, length = 20)

Similar question - why do you need this? 
Is this because of the last batch that may be smaller than the rest? Can't you just ignore the last few elements in the arrays instead of resizing? (resizing allocates a new array).

Reply:
Explained in email

---
In reply to: [427855107](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427855107) [](ancestors = 427855107)